### DISCUSSION TITLE
Oracles

### DISCUSSION THREAD

**Discussion by JaisPiyush** [JaisPiyush/arc-ip#3] [Ideas]
## Summary
Oracle sync off-chain data to on-chain, perform expensive functions, and cron jobs.

### Goal
1.  Data Feeding
    1. Cumulative earning of the protocol
    2. Cumulative balance data (Locked, Available, Pledged Collateral).
    3. Protocol raw byte power.
    4. Initial pledge collateral rate.
2. Miner Earning (every 7 days)
3. Penalty (7 days).
4. Oracle Gas uses (7 days).

|The address calling these methods must be registered as an oracle.|
|----------------------------------------------------------------|

#### Data Feeding
* In every fixed interval (number of blocks) static data is fed by the oracles.
* These data are deterministic and single value hence only requires consensus as proof of validation, not Merkle Trees.
```solidity

/// Submit the cumulative earning for the block
function submitCumulativeEarning(uint256 block, uint256 totalEarning) external;
function submitCumulativeBalance(uint256 block, uint256 totalLockedAmount, 
            uint256 totalAvailableBalance, uint256 totalPledgedCollateral) external;
function submitProtocolRawBytePower(uint256 block, uint256 protocolRawBytePower) external;
function submitCollateralRate(uint256 block, uint256 pledgeCollateralRatePerTB) external;
```

#### Miner Earning
* Miner earnings will be reported every 7 days using Merkle Tree proofs.
* Each node will have to claim the reported earnings through Merkle Tree proof and then withdraw the available balance into the StakePool.
* Amount of gas used from last withdrawal till now and some extra reward will be transferred to the Oracle address and deducted from the withdrawn balance.
* residual earnings will be divided between the node operator and user based on the share percentage of stake.
* Since the node operator has to take care of hardware, and the other cost factors for running the node, the node operator will always have a greater share of the earnings (earning shares of node = staked (1 + 10%)/ total stakes, while max. stake by the node operator can be 75%).

```solidity

/// submit the Merkle root for the earning of miners for 7 days
function submitEarningRoot(bytes32 merkleRoot) external;
/// claim the submitted earning after the establishment of the root for 7 days of earning
function submitEarningClaim(bytes32[] proof, uint256 earning) external;
````

#### Penalty
* Filecoin defines three types of penalty:
    - Broken contract
    - Sector Faults
    - Consensus Faults
* This section is yet to be solved

#### Oracle gas report
* Every action performed by the oracle must trigger an event with a timestamp for other oracles to track.
* The gas report must be proved using Merkle Tree proof because each oracle can have different methods performed during the time period.

**Comment by hcheng826** [comment]
I'm thinking we can increase the granularity for the data? 7 days sounds a bit long, or there's some reason for it?
https://filscan.io/tipset/chain
I can see the reward for each block on filecoin block explorer. Tried to look for some API calls for getting block reward and slashing value. And it's surprising to me that there isn't an easy way to get it. Or just I haven't found it. Also asked in the Filecoin Slack. Also found related issue in lotus repo: https://github.com/filecoin-project/lotus/issues/2376

**Reply by hcheng826** [reply]
But to me, I feel fixing this Oralce problem: syncing these block-related data to the FVM and make it available to the smart contract (Solidity). This issue itself can be a startup project. I will look around to see if there's any existing solution.

**Reply by JaisPiyush** [reply]
1.Syncing rewards in for loop can quickly outrun the upper gas limit of block. Instead Merkle Tree is used to update. Each new root requires consensus above 51% to be considered truth. Then each oracle will submit their proof to claim their node's reward. Claiming the reward is actually updating the smart contract data for that node.

2. Can be done for each day or block but uploading new merkle root and then claiming the rewards to update the smart contract can get bit tidious (for each block)

**Reply by JaisPiyush** [reply]
Chainlink will debut on fvm soon. But we need oracle for custom data (data for selected nodes) and claiming of rewards.

**Reply by hcheng826** [reply]
some updates from filecoin team
https://filecoinproject.slack.com/archives/CRK2LKYHW/p1678157310718409?thread_ts=1678115534.712109&cid=CRK2LKYHW

**Reply by JaisPiyush** [reply]
We have to work this out with insff. data, they have no implementation yet fr slashing

**Comment by JaisPiyush** [comment]
I don't think we need to submit Penalties if we are using fixed interest rate model

**Reply by hcheng826** [reply]
Just to confirm by "Submit" you mean we have a off-chain job that fetch the data and write the data to the on-chain states?

**Reply by JaisPiyush** [reply]
Yes,

**Comment by JaisPiyush** [comment]
Oracles requirement:
1. Submit off-chain data
2. Change the oracle address for a particular node
3. claim reward through cli
4. respond to challenges to prove the activeness on-chain.

**Reply by JaisPiyush** [reply]
Does the change of oracle address only via the current oracle address possess any risk? Should we also include the owner's permission or totally transfer the power to the owner?

**Comment by JaisPiyush** [comment]
Another development regarding Miner earning.

1. Miner earning is the total available balance in the miner account.

2. Only 90% of the available balance can be withdrawn from active node, 10% to provide for gas and few other functions.

3. Since all the earnings will be summed before distribution to the stakers. This comes with a disadvantage, the penalty scored by an individual miner will be shared with all the miners. So, the miner working ideally will have to bear the effects of notorious miners. To overcome this a complex earning distribution should be calculated with objective to fullfill miners share (including penalties and rewards) before disbursing to stakers.

5. For any node A, from the last snapshot block, earnings will be summed till the next snapshot block. The share of of node earning will be calculated according to the amount locked in the Stakepool. The residual of the amount after this will be distributed evenly amoung the stakers.

6. 90% amount should be taken after the entire summation. 

8. The merkle tree should be constructed with `(address, em uint256, es uint256)`, address of miner, em is earning claimable by the owner of thr node, es is earning claimable by stakers and em+es will be total claimable earning.

9. The es will be transferred to aFIL contract, from where the stakers can exchange their aFIL for FIL.
